sdk/tables/azure-data-tables/src/table_clients.cpp

```diff
@@ -566,30 +588,25 @@ Azure::Response<Models::DeleteEntityResult> TableClient::DeleteEntity(
 
 Azure::Response<Models::UpsertEntityResult> TableClient::UpsertEntity(
     Models::TableEntity const& tableEntity,
-    Models::UpsertKind upsertKind,
-    Core::Context const& context)
+    Models::UpdateMode updateMode,
+    Core::Context const& context) const
 {
-  try
-  {
-    switch (upsertKind)
-    {
-      case Models::UpsertKind::Merge: {
-        auto response = MergeEntity(tableEntity, context);
-        return Azure::Response<Models::UpsertEntityResult>(
-            Models::UpsertEntityResult(response.Value), std::move(response.RawResponse));
-      }
-      default: {
-        auto response = UpdateEntity(tableEntity, context);
-        return Azure::Response<Models::UpsertEntityResult>(
-            Models::UpsertEntityResult(response.Value), std::move(response.RawResponse));
-      }
+  Models::TableEntity tableEntityInternal = tableEntity;
+  // since the only diff between update and upsert is the lack of etag in upsert
+  // remove ETag from the entity
+  tableEntityInternal.SetETag("");
```

Comment:
Rather than manually copying and overriding the TableEntity state that the user passed in, we may want to consider not setting the IfMatch header in the case where no Etag is present, rather than setting it as the `*`.